### PR TITLE
Add mechanism to override drainability status

### PR DESCRIPTION
**Pull request by artemvmin**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Adds a mechanism for one rule to selectively override other rules. This allows for custom rules that override some outcomes (e.g. BlockDrain -> DrainOk) while not overriding others (e.g. SkipDrain).

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: cluster-autoscaler/simulator/drainability/rules/rules_test.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+				fakeRule{drainability.NewBlockedStatus(drain.NotEnoughPdb, nil)},
+			},
+			want: drainability.Status{
+				Outcome:   drainability.DrainOk,
+				Overrides: []drainability.OutcomeType{drainability.BlockDrain},
+			},
+		},
+		"override no match": {
+			rules: Rules{
+				fakeRule{drainability.Status{
```
Comment (x13n) [inline cluster-autoscaler/simulator/drainability/rules/rules_test.go:59]: nit: How about `drainability.NewDrainableStatus(drainability.OverridingOnly(drainability.SkipDrain))`? Or is it not really easier to read?

File: cluster-autoscaler/simulator/drainability/rules/rules.go [hunk 2]
```diff
@@ -86,10 +86,23 @@ func (rs Rules) Drainable(drainCtx *drainability.DrainContext, pod *apiv1.Pod) d
...
-		if d.Outcome != drainability.UndefinedOutcome {
-			return d
+		status := r.Drainable(drainCtx, pod)
+		if len(status.Overrides) > 0 {
+			candidates = append(candidates, status)
+			continue
+		}
+		for _, candidate := range candidates {
+			for _, override := range candidate.Overrides {
+				if status.Outcome == override {
```
Comment (x13n) [inline cluster-autoscaler/simulator/drainability/rules/rules.go:99]: Maybe worth logging the fact that override happened? At least at `V(5)`? This logic becomes non-trivial with this change, so some observability wouldn't hurt.

**Review by x13n — CHANGES_REQUESTED** [review]
Just minor comments, this looks like a reasonable mechanism.

File: cluster-autoscaler/simulator/drainability/rules/rules.go (hunk 2 above)
Comment (artemvmin) [reply cluster-autoscaler/simulator/drainability/rules/rules.go:102]: Great idea. Added a Name() to the Rule interface to make this log more useful.

**Pull request merged**